src/core/thread/csl_tx_scheduler.cpp

```diff
@@ -0,0 +1,271 @@
+    int16_t  diff          = static_cast<int16_t>(aChild.GetCslPhase() - period_offset);
+
+    if (diff > kCslFrameRequestAheadThreshold)
+    {
+        delay = static_cast<uint16_t>(aChild.GetCslPhase() - period_offset - kCslFrameRequestAheadThreshold) *
+                kUsPerTenSymbols;
+    }
+    else
+    {
+        delay = static_cast<uint16_t>(aChild.GetCslPeriod() + aChild.GetCslPhase() - period_offset -
+                                      kCslFrameRequestAheadThreshold) *
+                kUsPerTenSymbols;
+    }
+
+    return delay;
+}
+
+otError CslTxScheduler::HandleFrameRequest(Mac::TxFrame &aFrame)
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(mCslTxChild != nullptr, error = OT_ERROR_ABORT);
+
+    mBusy = true;
+    error = mCallbacks.PrepareFrameForChild(aFrame, mFrameContext, *mCslTxChild);
+    VerifyOrExit(error == OT_ERROR_NONE, OT_NOOP);
+
+    if (mCslTxChild->GetIndirectTxAttempts() > 0 || mCslTxChild->GetCslTxAttempts() > 0)
+    {
+        // For a re-transmission of an indirect frame to a sleepy
+        // child, we ensure to use the same frame counter, key id, and
+        // data sequence number as the previous attempt.
+
+        aFrame.SetIsARetransmission(true);
+        aFrame.SetSequence(mCslTxChild->GetIndirectDataSequenceNumber());
+
+        if (aFrame.GetSecurityEnabled())
+        {
+            aFrame.SetFrameCounter(mCslTxChild->GetIndirectFrameCounter());
+            aFrame.SetKeyId(mCslTxChild->GetIndirectKeyId());
+        }
+    }
+    else
+    {
+        aFrame.SetIsARetransmission(false);
+    }
+
+    aFrame.SetChannel(mCslTxChild->GetCslChannel());
+    aFrame.SetTxPhase(mCslTxChild->GetCslPhase());
+    aFrame.SetTxPeriod(mCslTxChild->GetCslPeriod());
+
+exit:
+    return error;
+}
+
+void CslTxScheduler::HandleSentFrame(const Mac::TxFrame &aFrame, otError aError)
+{
+    Child *child = mCslTxChild;
+
+    VerifyOrExit(child != nullptr, OT_NOOP); // The result is no longer interested by upper layer
+
+    mCslTxChild = nullptr;
+    mBusy       = false;
+
+    HandleSentFrame(aFrame, aError, *child);
+
+exit:
+    return;
+}
+
+void CslTxScheduler::HandleSentFrame(const Mac::TxFrame &aFrame, otError aError, Child &aChild)
+{
+    switch (aError)
+    {
+    case OT_ERROR_NONE:
+        aChild.ResetCslTxAttempts();
+        aChild.ResetIndirectTxAttempts();
+        break;
+    case OT_ERROR_NO_ACK:
+        aChild.IncrementCslTxAttempts();
+
+        otLogInfoMac("Csl tx to child %04x failed, attempt %d/%d", aChild.GetRloc16(), aChild.GetCslTxAttempts(),
+                     kMaxCslTriggeredTxAttempts);
+
+        // Fall through
+    case OT_ERROR_CHANNEL_ACCESS_FAILURE:
+    case OT_ERROR_ABORT:
+
+        // Even if Csl Tx attempts count reaches max, the message won't be
+        // dropped until indirect tx attempts count reaches max. So here it
+        // would set sequence number and schedule next csl tx.
+
```

Comment:
On `ABORT` the `aFrame` can be an empty frame.

I think we want to add a `if(!aFrame.IsEmpty())` over the entire block (from here up to but before `RescheduleCslTx()` statement below).